### PR TITLE
TRI-226/ Set webhooks configuration

### PR DESCRIPTION
**Pull request by Juan-Bianchi**
# Pull Request

## Description
Implement a security middleware to ensure that the created endpoints only receive data from Linear webhooks. Configure ngrok to establish a tunnel.

## Ticket Link
[TRI-226](https://linear.app/tricker-sirius/issue/TRI-226/set-webhooks-configuration)

## Steps to Reproduce
1. Run `docker copose down`
2. Create an ngrok account, set an static url and after that, add NGROK_AUTH_TOKEN and NGROK_STATIC_DOMAIN constants in .env file.
3. Run `docker compose build`
4. Run `docker compose up`

### CODE REVIEW

File: src/router/index.ts [hunk 1]
```diff
@@ -15,3 +17,7 @@ router.use('/issue', withAwsAuth, issueRouter);
 router.use('/projects', withAwsAuth, projectRouter);
 router.use('/roles', withAwsAuth, roleRouter);
 router.use('/issue-providers', withAwsAuth, issueProviderRouter);
+
+if (Constants.NODE_ENV === NodeEnv.DEV) {
+  router.use('/ngrok', ngrokHealthRouter);
```
Comment (luccar31) [inline src/router/index.ts]: This is gonna respond with OK code every time the project is running, independently if Ngrok is up and running or not

Comment (luccar31) [reply src/router/index.ts]: You can do a /health endpoint to check server health

File: src/domains/webhookSource/repository/webhookSource.repository.impl.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
+import { type WebhookSourceRepository } from '@domains/webhookSource/repository/webhookSource.repository';
+import { WebhookSourceDTO } from '@domains/webhookSource/dto';
+import { type PrismaClient } from '@prisma/client';
+
+export class WebhookSourceRepositoryImpl implements WebhookSourceRepository {
+  constructor(private readonly db: PrismaClient) {}
+
+  /**
+   * Retrieves a webhook source by its dynamic route.
```
Comment (luccar31) [inline src/domains/webhookSource/repository/webhookSource.repository.impl.ts]: Check doc

Comment (luccar31) [reply src/domains/webhookSource/repository/webhookSource.repository.impl.ts]: There's no dinamic route anymore

File: src/domains/health/ngrokHealth/ngrokHealth.controller.ts [hunk 3]
```diff
@@ -0,0 +1,7 @@
+import { type Request, type Response, Router } from 'express';
+import HttpStatus from 'http-status';
+export const ngrokHealthRouter = Router();
+
+ngrokHealthRouter.get('/', (req: Request, res: Response) => {
```
Comment (luccar31) [inline src/domains/health/ngrokHealth/ngrokHealth.controller.ts]: This can be called healthRouter

File: src/app.ts [hunk 4]
```diff
@@ -18,6 +18,15 @@ if (Constants.NODE_ENV === NodeEnv.DEV) {
   app.use(morgan('tiny')); // Log requests only in development environments
 }
 
+// Save raw body buffer before JSON parsing
+app.use(
+  express.json({
+    verify: (req: CustomRequest, res: Response, buf: Buffer) => {
+      req.rawBody = buf;
```
Comment (luccar31) [inline src/app.ts:25]: Why is this necessary? Provider documentation for this change.

**Review by luccar31 — CHANGES_REQUESTED** [review]

File: src/router/index.ts (hunk 1 above)
Comment (Juan-Bianchi) [reply src/router/index.ts]: solved

File: src/domains/health/ngrokHealth/ngrokHealth.controller.ts (hunk 3 above)
Comment (Juan-Bianchi) [reply src/domains/health/ngrokHealth/ngrokHealth.controller.ts]: solved

File: src/app.ts (hunk 4 above)
Comment (Juan-Bianchi) [reply src/app.ts:25]: Linear documentation [here](https://developers.linear.app/docs/graphql/webhooks)
image

File: src/domains/webhookSource/repository/webhookSource.repository.impl.ts (hunk 2 above)
Comment (Juan-Bianchi) [reply src/domains/webhookSource/repository/webhookSource.repository.impl.ts]: fixed

File: src/domains/webhook/middleware/middleware.linear.webhooks.ts [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+  }
+
+  // Decrypt the webhook source secret
+  const decryptedSecret: string = decryptData(webhookSource.secret);
+
+  // Verify signature using decrypted secret
+  const signature = verifyLinearSignature({ decryptedSecret, rawBody: req.rawBody });
+  if (signature !== req.headers['linear-signature']) {
+    throw new WebhookValidationException(400, 'Payload signature invalid.');
+  }
```
Comment (luccar31) [inline src/domains/webhook/middleware/middleware.linear.webhooks.ts]: Approch is okey, but this only works for Linear. My idea is to make a generic middleware for every provider, that implements the strategy pattern, so in order to add a new provider, we only need to create new Strategies.
In scense, it has to be something like (this)[https://refactoring.guru/design-patterns/strategy]. We can have multiple strategies. LinearVerifierStategy, JiraVerifierStategy, and so on. The stategy is selected in this case by the normalized name that comes into the `WebhookSourceDTO`. In this way, this middleware can be used for every possible provider and responsabilities are well separated.

**Review by luccar31 — CHANGES_REQUESTED** [review]

File: src/domains/webhook/middleware/middleware.linear.webhooks.ts (hunk 5 above)
Comment (luccar31) [reply src/domains/webhook/middleware/middleware.linear.webhooks.ts]:
```typescript
const providerWebhookVerificationRegistry: { [provider: string]: ProviderWebhookVerificationStrategy } = {};
function registerproviderWebhookVerification(provider: string, strategy: ProviderWebhookVerificationStrategy) {
  providerWebhookVerificationRegistry[provider] = strategy;
}

interface ProviderWebhookVerificationStrategy {
  verify: (request: ExpressRequest, source: WebhookSource) => Promise<boolean>
}

class LinearWebhookVerificationStrategy implements ProviderWebhookVerificationStrategy {
  verify(request: ExpressRequest, source: WebhookSource): Promise<boolean> {
    // linear verification logic
  }
}

class JiraWebhookVerificationStrategy implements ProviderWebhookVerificationStrategy {
  verify(request: ExpressRequest, source: WebhookSource): Promise<boolean> {
    // jira verification logic
  }
}


registerproviderWebhookVerification(new LinearWebhookVerificationStrategy())

providerWebhookVerificationRegistry['linear']
```
Take this as an example

File: src/domains/webhookSource/repository/webhookSource.repository.impl.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+  /**
+   * Retrieves a webhook source by its id.
+   * @param id The id route of the webhook source to retrieve.
+   * @returns A promise that resolves to a WebhookSourceDTO object if found, or null if not found.
+   */
+  async getById(id: string): Promise<WebhookSourceDTO | null> {
+    const webhookSource = await this.db.webhookSource.findUnique({
+      where: {
+        id,
+        deletedAt: { not: null },
```
Comment (luccar31) [inline src/domains/webhookSource/repository/webhookSource.repository.impl.ts]: deletedAt should be null. If otherwise, Prisma would try to find entities where deletedAt is NOT null, meaning, where entities are logically deleted.

**Review by luccar31 — CHANGES_REQUESTED** [review]

File: prisma/schema.prisma [hunk 7]
```diff
@@ -115,10 +115,16 @@ model WebhookOutcomeStatus {
...
 model WebhookSource {
-  id   String @id @default(dbgenerated("gen_random_uuid()")) @db.Uuid
-  name String
+  id                String    @id @default(dbgenerated("gen_random_uuid()")) @db.Uuid
+  providerWebhookId String?   @unique
+  secret            String?
+  issueProviderId   String    @db.Uuid
+  createdAt         DateTime  @default(now())
+  updatedAt         DateTime  @default(now())
+  deletedAt         DateTime?
```
Comment (luccar31) [inline prisma/schema.prisma]: This model needs to include the project relationship, the project the webhook is connected to

**Review by luccar31 — CHANGES_REQUESTED** [review]

File: prisma/schema.prisma (hunk 7 above)
Comment (Juan-Bianchi) [reply prisma/schema.prisma]: done

File: src/domains/webhookSource/repository/webhookSource.repository.impl.ts (hunk 6 above)
Comment (Juan-Bianchi) [reply src/domains/webhookSource/repository/webhookSource.repository.impl.ts]: done

**Review by luccar31 — APPROVED** [review]
